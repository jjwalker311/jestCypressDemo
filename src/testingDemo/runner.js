
/**
 * Runner function to call a given method for a range of test
 * @param  {Function} method
 * @param  {Object} tests
 */
export default function(method){

    /**
     * Returns function for a given method
     * @param  {Object} tests - see structure 
     */
    return function(tests){

        /**
         * Structure of TESTS Object
         * {
         *      TITLE_OF_TEST : {
         *          input : Array //arguments for method
         *          output : Any //expected return value for the method
         *          ?skip : Boolean
         *          ?only : Boolean
         *      }
         * }
         */
        Object.keys(tests).forEach(title =>{
            const {input, output, only, skip} = tests[title]
            //Run method with given args and expect output
            const testFn = () => {
                expect(method(...input)).toEqual(output)
            }
            
            if (only){
                test.only(title, testFn)
            } else if (skip){
                test.skip(title, testFn)
            } else {
                test(title, testFn)
            }
        })
    }
}   