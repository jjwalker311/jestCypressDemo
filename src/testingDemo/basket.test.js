import Basket from './basket'
import React from 'react'
import reducer from './stateReducer'

//Enzyme setup bleugh....
import Enzyme, {shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() })

import runner from './runner'

import {
    checkout as mockCheckout,
    addToBasket as mockAddToBasket,
    onCheckoutErr as mockOnCheckoutErr
} from '../mockServices/methods'


jest.mock('../mockServices/methods', () => {
    return {
        ...jest.requireActual('../mockServices/methods'),
        addToBasket : jest.fn(() => Promise.resolve({added : true})),
        checkout : jest.fn(),
        onCheckoutErr : jest.fn()
    }
})

const checkoutSuccess = {
    response : 200,
    next : 'myOrders'
}

const checkoutFailure = {
    response : 500
}

let userJourneyWrapper
let instance

const getQuantityDisplayed = () => userJourneyWrapper.find('p.quantity').text();

beforeAll(() => {
    mockCheckout.mockImplementation(() => Promise.resolve(checkoutSuccess))

    userJourneyWrapper  = shallow(<Basket/>)
    instance = userJourneyWrapper.instance()
})

afterEach(()=>{
    mockOnCheckoutErr.mockClear()
})

describe('Run sample test suite on Basket', ()=>{

    it('should render without blowing up', ()=>{
        expect(userJourneyWrapper.length).toBe(1)
        expect(userJourneyWrapper).toMatchSnapshot();

        expect(getQuantityDisplayed()).toMatch(/quantity of: 0/i)
        expect(userJourneyWrapper.find('Add').prop('onClick')).toBe((instance.handleAdd))
        expect(userJourneyWrapper.find('CheckoutButton').prop('onClick')).toBe(instance.handleCheckout)
    })

    it('should disable the checkout button by default', ()=>{
        const disabledValue = userJourneyWrapper.find('CheckoutButton').prop('disabled')
        expect(disabledValue).toBe(true)
    })

    it('should call the addToBasket method on handleAdd of add with correct args and update state', async ()=>{
        const event = {
            target : {
                id : 'someId'
            }
        }

        expect(getQuantityDisplayed()).toMatch(/quantity of: 0/i)

        await instance.handleAdd(event)

        //Quantity has been updated
        expect(getQuantityDisplayed()).toMatch(/quantity of: 1/i)
        expect(mockAddToBasket).toBeCalled()
        expect(mockAddToBasket).toBeCalledWith(true, event.target.id)

        await instance.handleAdd(event)
        expect(getQuantityDisplayed()).toMatch(/quantity of: 2/i)

        await instance.handleAdd(event)
        expect(getQuantityDisplayed()).toMatch(/quantity of: 3/i)

        await instance.handleAdd(event)
        expect(getQuantityDisplayed()).toMatch(/quantity of: 4/i)
    })

    it('should enable the checkout button by after handleAdd has been called', ()=>{
        const disabledValue = userJourneyWrapper.find('CheckoutButton').prop('disabled')
        expect(disabledValue).toBe(false)
    })

    it('should call onCheckoutErr if checkout response is not 200', async ()=>{
        expect(mockOnCheckoutErr).not.toBeCalled()
        mockCheckout.mockImplementationOnce(() => Promise.resolve(checkoutFailure))

        await instance.handleCheckout()

        expect(mockOnCheckoutErr).toBeCalled()
        expect(mockOnCheckoutErr).toBeCalledWith(`Response error: ${checkoutFailure.response}`)
    })

    it('should call onCheckoutErr if checkout throws an error', async ()=>{
        const err = 'someError'
        expect(mockOnCheckoutErr).not.toBeCalled()
        mockCheckout.mockImplementationOnce(() => Promise.reject(err))

        await instance.handleCheckout()

        expect(mockOnCheckoutErr).toBeCalled()
        expect(mockOnCheckoutErr).toBeCalledWith(err)
    })

    it('should not call error handler and succesfull checkout and redirect', async ()=>{
        const getRedirectionComponent = () => userJourneyWrapper.find('Redirect')
        const areWeRedirecting = () => {
            return getRedirectionComponent().length > 0
        }


        expect(areWeRedirecting()).toBe(false)

        await instance.handleCheckout()

        expect(mockOnCheckoutErr).not.toBeCalled()
        expect(areWeRedirecting()).toBe(true)
        expect(getRedirectionComponent().prop('to')).toBe(checkoutSuccess.next)
    })
})

describe('Run tests against state reducer', ()=>{
    
    let state = {
        quantity : 0
    }
    let action = {
        type : null,
        payload : null
    }

    it ('should return null for invalid type', ()=>{
        expect(reducer(state, action)).toEqual(null)

        action.type = 'something_unknown_not_in_enum'
        expect(reducer(state, action)).toEqual(null)
    })

    it('should return an error for an empty redirect payload and not with a valid payload', ()=>{
        action.type = 'redirect'

        const {redirectError : shouldBeTrue} = reducer(state, action)
        expect(shouldBeTrue).toEqual(true)

        action.payload = 'someRedirectActions'
        const {redirectError : shouldBeFalse} = reducer(state, action)
        expect(shouldBeFalse).toEqual(false)
    })

    it('should return redirect if valid payload sent', ()=>{
        const {redirect} = reducer(state, action)
        expect(redirect).toEqual(action.payload)
    })

    it('should increment quantity up to 10', ()=>{
        const expectedResults = [1,2,3,4,5,6,7,8,9,10]

        action.type = 'increment'

        expectedResults.forEach(result => {

            const updatedState = reducer(state, action)
            const {quantity, disableAddition} = updatedState

            expect(quantity).toBe(result)

            //Updating state for next test
            state = {
                ...state, ...updatedState
            }
        })
        
    })

    it('should disable the addition button when quantity is 10', ()=>{
        state.quantity = 0
        action.type

        let updatedState

        for(let i=1;i<=10;i++){
            updatedState = reducer(state, action)
            expect(updatedState.disableAddition).toEqual(i < 10 ? undefined : true)

            //Updating state for next assertion
            state = {
                ...state, ...updatedState
            }
        }
    })
})
