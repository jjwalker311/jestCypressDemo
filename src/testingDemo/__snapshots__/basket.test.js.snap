// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run sample test suite on Basket should render without blowing up 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Basket />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <React.Fragment>
        <p
          className="quantity"
          data-test="quantity"
        >
          Current Basket quantity of: 0
        </p>
        <Add
          onClick={[Function]}
        />
        <CheckoutButton
          disabled={true}
          onClick={[Function]}
        />
      </React.Fragment>,
      "className": "basket",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <p
            className="quantity"
            data-test="quantity"
          >
            Current Basket quantity of: 0
          </p>,
          <Add
            onClick={[Function]}
          />,
          <CheckoutButton
            disabled={true}
            onClick={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Current Basket quantity of: 0",
            "className": "quantity",
            "data-test": "quantity",
          },
          "ref": null,
          "rendered": "Current Basket quantity of: 0",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "disabled": undefined,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <React.Fragment>
          <p
            className="quantity"
            data-test="quantity"
          >
            Current Basket quantity of: 0
          </p>
          <Add
            onClick={[Function]}
          />
          <CheckoutButton
            disabled={true}
            onClick={[Function]}
          />
        </React.Fragment>,
        "className": "basket",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <p
              className="quantity"
              data-test="quantity"
            >
              Current Basket quantity of: 0
            </p>,
            <Add
              onClick={[Function]}
            />,
            <CheckoutButton
              disabled={true}
              onClick={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Current Basket quantity of: 0",
              "className": "quantity",
              "data-test": "quantity",
            },
            "ref": null,
            "rendered": "Current Basket quantity of: 0",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": undefined,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Symbol(react.fragment),
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
